src/lib.rs

```diff
@@ -474,6 +493,156 @@ impl XenControl {
         last_error!(self, max_gpfn)
     }
 
+    pub fn vcpu_getinfo(&self, domid: u32, vcpu: u32) -> Result<xc_vcpuinfo_t, XcError> {
+        debug!("vcpu_getinfo");
+        let xc = self.handle.as_ptr();
+        let mut vcpu_info = unsafe { mem::zeroed() };
+
+        (self.libxenctrl.clear_last_error)(xc);
+        (self.libxenctrl.vcpu_getinfo)(xc, domid, vcpu, &mut vcpu_info);
+
+        last_error!(self, vcpu_info)
+    }
+
+    pub fn physinfo(&self) -> Result<xc_physinfo_t, XcError> {
+        debug!("physinfo");
+        let xc = self.handle.as_ptr();
+        let mut physinfo = unsafe { mem::zeroed() };
+
+        (self.libxenctrl.clear_last_error)(xc);
+        (self.libxenctrl.physinfo)(xc, &mut physinfo);
+
+        last_error!(self, physinfo)
+    }
+
+    pub fn get_cpuinfo<'a>(
+        &'_ self,
+        infos: &'a mut [mem::MaybeUninit<xc_cpuinfo_t>],
+    ) -> Result<&'a [xc_cpuinfo_t], XcError> {
+        debug!("get_cpuinfo");
+        let xc = self.handle.as_ptr();
+        let mut nr_cpu = 0i32;
+
+        (self.libxenctrl.clear_last_error)(xc);
+        (self.libxenctrl.get_cpuinfo)(xc, infos.len() as i32, infos.as_mut_ptr() as _, &mut nr_cpu);
+
+        assert!((nr_cpu as usize) <= infos.len());
+
+        last_error!(
+            self,
+            // Reinterpret infos slice as [xc_cpuinfo_t], which is valid as :
+            //  - MaybeUninit is repr(transparent)
+            //  - infos outlives its return value due to explicit borrow
+            //  - nr_cpu <= infos.len()
+            slice::from_raw_parts(infos.as_ptr() as _, nr_cpu as usize)
+        )
+    }
+
+    pub fn get_cpufreq_avg(&self, cpuid: u32) -> Result<u32, XcError> {
+        debug!("get_cpufreq_avg");
+        let xc = self.handle.as_ptr();
+        let mut freq: c_int = 0;
+
+        (self.libxenctrl.clear_last_error)(xc);
+        (self.libxenctrl.get_cpufreq_avgfreq)(xc, cpuid as c_int, &mut freq);
+
+        last_error!(self, freq as _)
+    }
+
+    pub fn get_pxstat(&self, cpuid: u32, px_stat: &mut PxStat) -> Result<(), XcError> {
```

Comment:
This is not idiomatic Rust.
You should return the struct instead of modifying a mutable borrow.

Reply:
PxStat/CxStat contains several vectors that may be quite large in some cases (like transition_table which is max_px*max_px large) and this function may be called quite repetitively (once per core per update).

So my idea was to reuse the structure to prevent allocating over and over vectors of the same size.

Reply:
Ok, now I understand.
But you should add a comment to explain that choice for future reference.